gax/src/main/java/com/google/api/gax/rpc/FirstElementResponseObserver.java

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright 2017, Google LLC All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.rpc;
+
+import com.google.api.core.AbstractApiFuture;
+import com.google.api.core.ApiFuture;
+
+/**
+ * A {@link ResponseObserver} that wraps a future. The future will resolved upon receiving the first
+ * element or exhausting the streaming.
+ *
+ * <p>Package-private for internal use.
+ *
+ * @param <ResponseT> The type of the element in the stream.
+ */
+class FirstElementResponseObserver<ResponseT> implements ResponseObserver<ResponseT> {
+  private final MyFuture future = new MyFuture();
+  private StreamController controller;
+
+  @Override
+  public void onStart(StreamController controller) {
+    this.controller = controller;
+
+    controller.disableAutoInboundFlowControl();
+    controller.request(1);
+  }
+
+  @Override
+  public void onResponse(ResponseT response) {
+    future.set(response);
+    controller.cancel();
+  }
+
+  @Override
+  public void onError(Throwable t) {
+    future.setException(t);
+  }
+
+  @Override
+  public void onComplete() {
+    future.set(null);
+  }
+
+  ApiFuture<ResponseT> getFuture() {
+    return future;
+  }
+
+  /**
+   * Simple implementation of a future to prematurely interrupt the RPC before the first element is
+   * received.
```

Comment:
This comment doesn't seem to match the implementation - looking at the implementation, it seems that this future's role is to cancel a stream when it is cancelled. It looks like you could have a shared future with `SpoolingResponseObserver`, which has a `StreamController` passed to it. 

Looking more carefully: isn't it possible for `controller` to be null in some circumstances? Maybe it won't happen since FirstElementResponseObserver is package-private and we are in control of the code paths... Hmm, I do see a note in `SpoolingResponseObserver` to this effect, but not in the present class.

Reply:
I added some docs to clarify the situation. Let me know if it clarifies the situation or if I can improve the docs or if I'm missing something